framework/contracts/native/ibc-client/src/ibc.rs

```diff
@@ -1,526 +1,102 @@
-            Some(mut acct) => {
-                // set the account the first time
-                if acct.remote_addr.is_none() {
-                    acct.remote_addr = Some(account);
-                }
-                Ok(acct)
+    // 2. From here on, we can trust the message that we are receiving
+
+    let callback_msg: IbcClientCallback = from_binary(&callback.initiator_msg)?;
+
+    match callback_msg {
+        IbcClientCallback::WhoAmI {} => {
+            // This response is used to store the Counterparty proxy address (this is used to whitelist the address on the host side)
+            if let Callback::Execute(Ok(response)) = &callback.result {
+                IBC_INFRA.update(deps.storage, &host_chain, |c| match c {
+                    None => Err(IbcClientError::UnregisteredChain(host_chain.to_string())),
+                    Some(mut counterpart) => {
+                        counterpart.remote_proxy = Some(response.executed_by.clone());
+                        Ok(counterpart)
+                    }
+                })?;
+            } else {
+                return Err(IbcClientError::IbcFailed(callback));
             }
-            None => Err(IbcClientError::UnregisteredChannel(channel_id.clone())),
-        }
-    })?;
-
-    Ok(IbcBasicResponse::new().add_attribute("action", "acknowledge_register"))
-}
-
-// receive PacketMsg::Balances response
-fn acknowledge_balances(
-    deps: DepsMut,
-    env: Env,
-    channel_id: String,
-    account_id: AccountId,
-    ack: StdAck,
-) -> Result<IbcBasicResponse, IbcClientError> {
-    // ignore errors (but mention in log)
-    let BalancesResponse { account, balances } = match ack {
-        StdAck::Result(res) => from_slice(&res)?,
-        StdAck::Error(e) => {
-            return Ok(IbcBasicResponse::new()
-                .add_attribute("action", "acknowledge_balances")
-                .add_attribute("error", e))
+            Ok(IbcClientResponse::action("register_remote_proxy")
+                .add_attribute("chain", host_chain.to_string()))
         }
-    };
-
-    ACCOUNTS.update(deps.storage, (&channel_id, account_id), |acct| match acct {
-        Some(acct) => {
-            if let Some(old) = acct.remote_addr {
-                if old != account {
-                    return Err(IbcClientError::RemoteAccountChanged { old, addr: account });
-                }
+        IbcClientCallback::CreateAccount { account_id } => {
+            // We need to get the address of the remote proxy from the account creation response
+            if let Callback::Execute(Ok(response)) = &callback.result {
+                let account_creation_result = response.result[0].clone();
+
+                let wasm_event = account_creation_result
+                    .events
+                    .into_iter()
+                    .find(|e| e.ty == "wasm")
+                    .ok_or(IbcClientError::IbcFailed(callback.clone()))?;
+                let remote_proxy_address = wasm_event
```

Comment:
Hmmmmm this seems a bit vulnerable but I don't see a good way to do it another way.